### PR TITLE
[SPARK-31026] [SPARK-31060] [SQL] [test-hive1.2] Parquet predicate pushdown on columns with dots

### PR DESCRIPTION
**Pull request by dbtsai**
### What changes were proposed in this pull request?
Parquet predicate pushdown on columns with dots is disabled in [SPARK-20364](https://jira.apache.org/jira/browse/SPARK-20364) due to the limitation of Parquet APIs.

A new set of APIs is purposed in [PARQUET-1809](https://jira.apache.org/jira/browse/PARQUET-1809) to generalize the support for both cols containing `dot` and nested cols.

This PR implements a new Parquet filter APIs that supports both column names containing `dot` and nested columns. We will remove those code from Spark codebase once we upgrade to a new release of Parquet that contains this implementation. 

### Why are the changes needed?
Many tables in production are using `dot` as part of the column names, and without predicate pushdown on those columns, the performance is suffering.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
Existing tests and one new test.

### CODE REVIEW

**Comment by dbtsai** [comment]
This depends on https://github.com/apache/spark/pull/27778 . Once the other one is merged, I will rebase against master. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -55,10 +55,12 @@ class ParquetFilters(
     // and it does not support to create filters for them.
     val primitiveFields =
     schema.getFields.asScala.filter(_.isPrimitive).map(_.asPrimitiveType()).map { f =>
-      f.getName -> ParquetField(f.getName,
+      import org.apache.spark.sql.connector.catalog.CatalogV2Implicits._
+      quote(f.getName) -> ParquetField(f.getName,
         ParquetSchemaType(f.getOriginalType,
           f.getPrimitiveTypeName, f.getTypeLength, f.getDecimalMetadata))
     }
+    val b = 10
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Maybe, a test left-over? Shall we remove this?

Comment (dbtsai) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: Oops. Thanks.

File: sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.parquet.filter2.predicate.Operators.BooleanColumn;
+import org.apache.parquet.filter2.predicate.Operators.DoubleColumn;
+import org.apache.parquet.filter2.predicate.Operators.FloatColumn;
+import org.apache.parquet.filter2.predicate.Operators.IntColumn;
+import org.apache.parquet.filter2.predicate.Operators.LongColumn;
+
+/**
+ * TODO (PARQUET-1809): This is a temporary workaround; it is intended to be moved to Parquet.
+ */
+public final class SparkFilterApi {
```
Comment (HyukjinKwon) [inline sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java]: I remember @rdblue didn't like this way. Am I correct?

File: sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.parquet.hadoop.metadata.ColumnPath;
+import org.apache.parquet.filter2.predicate.Operators.BinaryColumn;
+import org.apache.parquet.filter2.predicate.Operators.BooleanColumn;
+import org.apache.parquet.filter2.predicate.Operators.DoubleColumn;
+import org.apache.parquet.filter2.predicate.Operators.FloatColumn;
+import org.apache.parquet.filter2.predicate.Operators.IntColumn;
+import org.apache.parquet.filter2.predicate.Operators.LongColumn;
+
+/**
+ * TODO (PARQUET-1809): This is a temporary workaround; it is intended to be moved to Parquet.
```
Comment (HyukjinKwon) [inline sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java]: Maybe I am remembering wrongly but I initially tried to allow filters with dots with the similar approach here (https://github.com/apache/spark/pull/18000). It was suggested simply to disable it so I did it, and @rdblue didn't like it either. Am I correct?

**Comment by dongjoon-hyun** [comment]
Please rebase to the master because the related sub-PR is merged now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala [hunk 4]
```diff
@@ -1121,6 +1121,29 @@ abstract class ParquetFilterSuite extends QueryTest with ParquetTest with Shared
         }
       }
     }
+
+    withSQLConf(
+      // Makes sure disabling 'spark.sql.parquet.recordFilter' still enables
+      // row group level filtering.
+      SQLConf.PARQUET_RECORD_FILTER_ENABLED.key -> "false",
+      SQLConf.PARQUET_FILTER_PUSHDOWN_ENABLED.key -> "true",
+      SQLConf.PARQUET_VECTORIZED_READER_ENABLED.key -> "false") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Can we test both vectorized and non-vectorized reader?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala [hunk 5]
```diff
@@ -122,10 +122,10 @@ private[sql] object CatalogV2Implicits {
           s"$quoted is not a valid TableIdentifier as it has more than 2 name parts.")
     }
 
-    def quoted: String = parts.map(quote).mkString(".")
+    def quoted: String = parts.map(quoteIfNeeded).mkString(".")
   }
 
-  def quote(part: String): String = {
+  def quoteIfNeeded(part: String): String = {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala]: Could you make another PR for this renaming first?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala]: +1

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 6]
```diff
@@ -503,38 +502,38 @@ class ParquetFilters(
     predicate match {
       case sources.IsNull(name) if canMakeFilterOn(name, null) =>
         makeEq.lift(nameToParquetField(name).fieldType)
-          .map(_(nameToParquetField(name).fieldName, null))
+          .map(_(Array(nameToParquetField(name).fieldName), null))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: to confirm, this PR doesn't support nested fields yet, right?

File: sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java (hunk 3 above)
Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/parquet/filter2/predicate/SparkFilterApi.java]: This looks more useful now as it can not only support column name with dots, but also nested fields.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala (hunk 6 above)
Comment (dbtsai) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala]: This PR doesn't support nested fields yet, but it's a one step forward.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala (hunk 5 above)
Comment (dbtsai) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogV2Implicits.scala]: Here is PR for renaming and consolidating two quoteIfNeeded implementations. https://github.com/apache/spark/pull/27814

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: +1, we can merge this to be code block above, which is inside a `Seq(true, false).foreach { vectorized =>`

**Review by cloud-fan — APPROVED** [review]
LGTM except one comment in test, thanks for cleaning this up and fix it!

**Comment by dbtsai** [comment]
This depends on https://github.com/apache/spark/pull/27817/files

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala (hunk 4 above)
Comment (dbtsai) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala]: Done. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala [hunk 7]
```diff
@@ -32,6 +32,7 @@ import org.apache.spark.annotation.{Evolving, Stable}
 sealed abstract class Filter {
   /**
    * List of columns that are referenced by this filter.
+   * Note that, if a column contains `dots` in name, it will be quoted to avoid confusion.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/sources/filters.scala]: This one is actually a pretty breaking change. Not all implementations of the data sources will have the syntax to handle backquotes - there are so many non-DBMS implementations outside like elasticsearch, mongodb, etc. which I see relevant tickets in Spark JIRAs time to time.

In particular, this is a stable API. Can we update the migration guide at the very least?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFiltersBase.scala [hunk 8]
```diff
@@ -45,4 +47,11 @@ trait OrcFiltersBase {
     case _: AtomicType => true
     case _ => false
   }
+
+  /**
+   * The key of the dataTypeMap will be quoted if it contains `dots`.
+   */
+  protected[sql] def quotedDataTypeMap(schema: StructType): Map[String, DataType] = {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcFiltersBase.scala]: nit: `protected[sql]` -> `protected[orc]`

**Comment by dbtsai** [comment]
Closing it and merging with https://github.com/apache/spark/pull/27728/files
